src/hotspot/share/memory/arena.hpp

```diff
@@ -205,4 +205,8 @@ class Arena : public CHeapObjBase {
 #define NEW_ARENA_OBJ(arena, type) \
   NEW_ARENA_ARRAY(arena, type, 1)
 
+#ifdef ASSERT
+bool Arena_contains(const Arena* arena, const void* ptr);
```

Comment:
@kimbarrett the reason was that I need to call this from the hpp file, and I encountered some circular dependency I did could not resolve. So I needed to move something off to the cpp files. Either I put it in arena.cpp, or in growableArray.cpp. But If I put things off to growableArray.cpp from the GrowableArray class, then it will not easily instantiate the templates, so that is not possible then. Hence I have to put it into arena.cpp

Reply:
I don't think a global API is warranted to support that local debug-only
implementation detail.

The inclusion problem arises because the PR eliminates GrowableArrayMetadata 
(a non-templated class), thereby forcing the init_checks helper to be moved to
GrowableArray (a class template).  That forced moving the implementation from
the .cpp to the .hpp.

One solution might be to move the new version of init_checks to
growableArray.inline.hpp. After all, breaking circularities is one of the
reasons one might have an inline.hpp file. However, that file doesn't
currently exist, and I think introducing it would have too much fannout. And
it might not even work without even more effort; there might be .hpp files
that contain allocations of GrowableArray.

Much simpler, and I think probably better, is to keep GrowableArrayMetadata
(perhaps under a different name - GrowableArrayArenaHolder?), now holding the
allocation arena and providing init_checks (and anything else that seems
appropriate), with init_checks defined similarly to the current definition in
the .cpp file.